### PR TITLE
Add support for Render

### PR DESCRIPTION
**Pull request by voxbono**
Add a new action called Render which will add support for view_engines and server-rendered HMTL pages. I have added handlebars to dev-dependencies to show an example.

I wasn't sure how you wanted the routes to be defined, so moved the old '/' to '/json' and added a new route for '/' that shows a server-rendered HTML page.

I have not provided any tests for this as I struggled to get the test framework running. (some local issues).

### CODE REVIEW

**Comment by Avaq** [comment]
I just opened #29 with the maintenance work I mentioned. It brings the approach from [Fluture Template](https://github.com/fluture-js/fluture-template) into this project, which allows it to be consistent with all the other Fluture projects.

If you like to give it a try, you can rebase your PR on mine, and then I can review it. If you feel like the change is a bit daunting, I can review your PR in its current form, merge it, and then rebase my own PR.

File: example/views/index.hbs
Comment (Avaq) [inline example/views/index.hbs:10]: Please add a newline to the end of this file, or [it won't be universally acknowledged as a text file](https://unix.stackexchange.com/questions/18743/whats-the-point-in-adding-a-new-line-to-the-end-of-a-file/18789#18789).

**Comment by voxbono** [comment]
my changes are quite small. So if you can merge maintenance in to master when you are done, I can do it from there. It doesn't look too complicated.

File: index.js
Comment (Avaq) [inline index.js:102]: Could you move this below the documentation for `Json`?

File: index.js
Comment (Avaq) [inline index.js:109]: Could you move this below the `Json` constructor definition?

Also, I think it would be good to add a `code` parameter. That would allow users to use this to render custom 404 pages, for example.

File: index.js
Comment (Avaq) [inline index.js:151]: Also move below `Json` and add the `code` parameter.

**Comment by voxbono** [comment]
Tried to resolve all the issues

**Comment by voxbono** [comment]
> I just opened #29 with the maintenance work I mentioned. It brings the approach from [Fluture Template](https://github.com/fluture-js/fluture-template) into this project, which allows it to be consistent with all the other Fluture projects.
> 
> If you like to give it a try, you can rebase your PR on mine, and then I can review it. If you feel like the change is a bit daunting, I can review your PR in its current form, merge it, and then rebase my own PR.

I'm haven't been working that much in open source projects. So not sure how to rebase another pull request into my branch. I will be happy to do it, but I'm afraid to mess it up and just create more work for you.

File: index.js
Comment (Avaq) [inline index.js:88]: You forgot to update the documentation after adding the new parameter. How about:

*Indicates a response to be rendered using a template. The first argument will be the response status code, the second is the path to the template file, and the third is the data to inject into the template. This uses Express' `render` method under the hood, so you can configure it globally with `app.set ('view engine', engine)` and `app.set ('views', path)`.*

**Comment by Avaq** [comment]
> So not sure how to rebase another pull request into my branch

I'll guide you through it.

1. First address my last point of feedback and commit it to your branch.
2. Now we will squash your commits into a single commit. This will make your commit history a bit cleaner, but it will also make the next step easier:
    ```sh
    # Add my copy of the repository as another remote
    git remote add upstream https://github.com/fluture-js/fluture-express.git

    # Fetch the new remote
    git fetch --all

    # Rebase on the master branch interactively. This is what will allow you to "squash" your commits.
    git rebase --interactive upstream/master
    # Now your editor will open with a commit on every line
    # Replace the word "pick" on every line except the first one with the word "fixup"
    # Save and exit

    # Force push to your branch. The force is needed because we just rewrote history.
    git push origin render --force
    ```
3. Now you can do the rebase onto my other branch.
    ```sh
    # We do another rebase, but this time on my WIP branch.
    git rebase upstream/avaq/template
    # At this point, you'll probably get a bunch of conflicts, just like when you're merging.
    # Resolve each conflict.

    # After the conflicts are resolved, you need to stage the fixed files
    git add .

    # Finish the rebase.
    git rebase --continue

    # We rewrote history again.
    git push origin render --force
    ```

After the rebase, you can also try running the tests locally, because we're using a new testing framework. Maybe the new one will work better for you. :)
If it does work, you can add some tests for the new feature you've created.

**Comment by voxbono** [comment]
A few comments:
I'm not happy with the test for Render. I wasn't sure how to test it. I something that doesn't fail, but still I think the test could be better.

When running npm i, the repo is reporting 46 vulnerabilities (34 low, 2 moderate, 9 high, 1 critical). I can't run npm audit, as the repo is missing package-lock. So not sure where the vulnerabilities are.

I also struggled a lot with the usage of import instead of require. But using node 14 fixed that issue for me.

**Comment by Avaq** [comment]
You managed the rebase! :D
Excellent. I just merged #29, so you can rebase on master one last time, and push to clean up the history. You can do it like this:

```sh
# Fetch the new master branch
git fetch --all

# Rebase into the new master
git rebase --interactive upstream/master
# In the editor that opens, simply remove all lines that are not referring to commits that you made.
# So you'll only keep the last two lines.
# Then save and exit. There should be no conflicts.

# We rewrote history one last time
git push origin render --force
```

File: example/actions/session.js
Comment (Avaq) [inline example/actions/session.js:8]: Maybe we can split this into several statements to improve legibility? I normally would use a `pipe` here, but it might not be worth adding that.

```suggestion
module.exports = (req, locals) => {
  const session = {id: req.headers['x-authenticated-user']};
  const newLocals = Object.assign ({session}, locals);
  return Future.resolve (Next (newLocals));
};
```

File: example/actions/welcome.js
Comment (Avaq) [inline example/actions/welcome.js:10]: We could take a similar approach here

```suggestion
module.exports = (req, locals) => {
  const user = locals.session.id ? `user ${locals.session.id}` : 'stranger';
  return Future.resolve (Render (200) ('index') ({user}))
};
```

File: example/actions/welcomeJson.js
Comment (Avaq) [inline example/actions/welcomeJson.js:9]: ... and here

```suggestion
module.exports = (req, locals) => {
  const welcome = locals.session.id ? `user ${locals.session.id}` : 'stranger';
  Future.resolve (Json (200) ({welcome}));
};
```

File: test/index.js
Comment (Avaq) [inline test/index.js:129]: Could we add an assertion to check if `render` got called correctly? eg:

```suggestion
  eq (mockRes.status.args, [[200]]);
  eq (mockRes.render.args, [['index', {user: 'hello'}]]);
```

File: example/actions/image.js
Comment (voxbono) [reply example/actions/image.js:2]: sorry, forgot to add the back. I had some linting issues at an earlier stage. Will add them back again.

Comment (voxbono) [reply example/actions/image.js:2]: By the way, is 'use strict'; helping us here. Linter is complaining that it is unnecessary inside of modules. And since I have autofix on save on in VSCode, it just gets removed. I normally don't use strict mode in my own projects. Just want to know why you want it there.

File: test/index.js
Comment (voxbono) [reply test/index.js:129]: Yup, this seemed to work. Committing.

File: README.md
Comment (voxbono) [inline README.md:66]: I made the line numbers point to the comment for the function in index.js. Also, wasn't sure if blob/v3.1.0 was correct.

Comment (davidchambers) [reply README.md:66]: __README.md__ is automatically updated at release time, so you should revert your changes to this file. Read about [Transcribe][1] if you're curious. :)


[1]: https://github.com/plaid/transcribe

Comment (Avaq) [reply README.md:66]: @voxbono I think there's been a misunderstanding. I asked for `example/README.md` to be updated, but you probably thought I meant `/README.md`. What @davidchambers said is right: `/README.md` is generated from code comments, and shouldn't be manually modified. It will be automatically built when I release the new version, with line numbers and all.

**Comment by Avaq** [comment]
@voxbono if you remove your last commit, and create a new commit that updates `example/README.md`, then this should be good to go.

**Comment by voxbono** [comment]
@Avaq So sorry for being so slow in this commit. Been very hectic lately. I finally managed to do it. Please verify that everything is ok.

File: example/actions/image.js
Comment (voxbono) [reply example/actions/image.js:2]: Hmm... won't this require node version > 14? I'm just afraid people will get errors in their projects. One thing is that the latest release of Node is > 14. But they still recommend 12.x.x. 
This is not an issue for someone messing around in a private project, but when you work for a company, you are sometimes not allowed to use the latest version. But I guess, you have already made Fluture dependent on version 14 now, right? So I guess it doesn't matter...

I will let this one be your call! :)

Comment (Avaq) [reply example/actions/image.js:2]: > already made Fluture dependent on version 14

Not *dependent on*, but *compatible with*. Older versions of Node use the commonjs files by default.

> I will let this one be your call! :)

Sure. If that's okay I will make the desired changes and push them to your branch before merging.

**Review by Avaq — APPROVED** [review]

**Pull request merged**
